File: yandex_back_prep_python/leetcode_python/303. Range Sum Query - Immutable.py
class NumArray:
    def __init__(self, nums: List[int]):
        self.nums = nums
        print(nums)
        self.pref_sums = [0] * (len(nums) + 1)
        for i in range(1, len(nums) + 1):
            self.pref_sums[i] = self.pref_sums[i - 1] + nums[i - 1]

    def sumRange(self, left: int, right: int) -> int:
        return self.pref_sums[right + 1] - self.pref_sums[left]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: yandex_back_prep_cpp/303. Range Sum Query - Immutable.cpp
class NumArray
{
public:
    NumArray(vector<int> &nums)
    {
        this->prefix = std::vector<int>(nums.size() + 1, 0);

        for (int i = 1; i < prefix.size(); i++)
        {
            prefix[i] = prefix[i - 1] + nums[i - 1];
        }
    }

    int sumRange(int left, int right)
    {
        return this->prefix[right + 1] - this->prefix[left];
    }

    std::vector<int> prefix;
};